File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;

class Solution {
    private int n, m;
    private int[] amounts;
    
    public int solution(int[][] land) {
        int answer = 0;
        n = land.length;
        m = land[0].length;
        
        int number = 2;
        amounts = new int[(n * m) + 2];
        
        for(int i=0 ; i<n ; i++){
            for(int j=0 ; j<m ; j++){
                if(land[i][j] != 1)
                    continue;
                bfs(i, j, land, number);
                number++;
            }   
        }
        
        for(int j=0 ; j<m ; j++){
            boolean[] visit = new boolean[number];
            int total = 0;
            
            for(int i=0 ; i<n ; i++){
                if(visit[land[i][j]] || land[i][j] == 0)
                    continue;
                
                total += amounts[land[i][j]];
                visit[land[i][j]] = true;
            }
            answer = Math.max(answer, total);
        }
        
        return answer;
    }
    
    private void bfs(int y, int x, int[][] land, int number){
        Queue<int[]> que = new LinkedList<>();
        que.add(new int[]{y, x});
        
        land[y][x] = number;
        
        final int[] dy = {0,0,1,-1}, dx = {1,-1,0,0};
        
        int cnt = 0;
        while(!que.isEmpty()){
            int[] cur = que.poll();
            int curY = cur[0], curX = cur[1];
            
            cnt++;
            
            for(int i=0 ; i<4 ; i++){
                int nextY = curY + dy[i], nextX = curX + dx[i];
                
                if(0 > nextY || nextY >= n || 0 > nextX || nextX >= m 
                   || land[nextY][nextX] != 1)
                    continue;
                
                que.add(new int[]{nextY, nextX});
                land[nextY][nextX] = number;
            }
        }
        amounts[number] = cnt;
    }
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import deque

def bfs(land, number, y, x, n, m):
    que = deque()
    que.append((y, x))
    
    dy, dx = (0,0,1,-1), (1,-1,0,0)
    
    cnt = 1
    land[y][x] = number
    
    while que:
        cur_y, cur_x = que.pop()
        
        for i in range(4):
            next_y, next_x = cur_y + dy[i], cur_x + dx[i]
            
            if not (0 <= next_y < n and 0 <= next_x < m) \
            or land[next_y][next_x] != 1:
                continue
                
            cnt += 1
            que.append((next_y, next_x))
            land[next_y][next_x] = number
            
    return cnt


def solution(land):
    answer = 0
    
    n = len(land)
    m = len(land[0])
    oil = [0] * (n * m)
    number = 2
    
    for y in range(n):
        for x in range(m):
            if land[y][x] != 1:
                continue
            
            oil[number] = bfs(land, number, y, x, n, m)
            number += 1
            
    for x in range(m):
        total = 0
        visit = {}
        for y in range(n):
            if land[y][x] == 0 or visit.get(land[y][x]):
                continue
            
            total += oil[land[y][x]]
            visit[land[y][x]] = True
            
        answer = max(answer, total)
    
        
    return answer